client/src/main/scala/Effects.scala

```diff
@@ -63,4 +65,19 @@ object Effects {
       case _ ⇒ noop
     }
 
+  def validateState(state: List[ClientExercise]): List[ClientExercise] =
+    state.foldLeft(List.empty[ClientExercise]) {
+      (acc: List[ClientExercise], exercise: ClientExercise) =>
+        DomHandler
+          .findExerciseByMethod(exercise.method)
+          .map(DomHandler.inputsInExercise)
+          .fold(acc) { inputs =>
+            if (exercise.arguments.size <= inputs.size) {
```

Comment:
Is it not better to always invalidate the exercise, I mean when:
`if (exercise.arguments.size == inputs.size) {
acc :+ exercise
} else {
acc :+ exercise.copy(arguments = Seq.empty, state = Unsolved)
}
`
I understand that you cannot better detect if the exercise was changed or not, but if it was the change can be anything, even the change of the params order, thus is better to mark it as Unresolved.

I dunno how `Unsolved` is interpreted in UI.

IMO exercise should be correctly set as `Unsloved` when there is a change. Then you can choose if the arguments will be preserved, always, i.e. striped when they are too long or always cleared.

Reply:
Otherwise it solves my issue.

Reply:
Yeah the downside of this is that we're storing in the DB the user progress, which basically carries the following data: which exercise was, which fields were completed, and what was the state of the exercise in that moment (completed successfully or not).

In any case you're right. I'm going to extend this fix to all exercises that have changed the number of arguments from the user progress (even if they've added stuff). We don't have a way to know if the order or expected values changed, but I think it's sensible. I still think that invalidating the fields is the safest option here, but maybe other members of the team can input some suggestions about this.

Thanks!